nuxeo-features/nuxeo-platform-comment/nuxeo-platform-comment/src/main/java/org/nuxeo/ecm/platform/comment/impl/CommentsMigrator.java

```diff
@@ -59,64 +74,161 @@
+
+        if (!rootCommentsFolderIds.isEmpty()) {
+            String query = buildAllCommentsQuery(rootCommentsFolderIds, HIDDEN_FOLDER_TYPE);
+            List<String> timestampCommentFoldersIds = session.queryProjection(query, 0, 0)
+                                                             .stream()
+                                                             .map(entry -> (String) entry.get(ECM_UUID))
+                                                             .collect(Collectors.toList());
+            parentIds.addAll(rootCommentsFolderIds);
+            parentIds.addAll(timestampCommentFoldersIds);
+        }
+        return parentIds;
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected void migrateSessionRelationToProperty(CoreSession session, MigrationContext migrationContext) {
         CommentService commentComponent = Framework.getService(CommentService.class);
         CommentServiceConfig commentServiceConfig = commentComponent.getConfig();
         if (commentServiceConfig != null) {
             RelationManager relationManager = Framework.getService(RelationManager.class);
             Graph graph = relationManager.getGraph(commentServiceConfig.graphName, session);
             List<Statement> statements = graph.getStatements();
-            checkShutdownRequested();
+            checkShutdownRequested(migrationContext);
 
-            processBatched(BATCH_SIZE, statements,
-                    statement -> migrateComments(session, relationManager, commentServiceConfig, statement),
-                    "Migrating comments");
-            reportProgress("Done", statements.size(), statements.size());
+            processBatched(migrationContext, BATCH_SIZE, statements,
+                    statement -> migrateCommentsFromRelationToProperty(session, relationManager, commentServiceConfig,
+                            statement),
+                    "Migrating comments from Relation to Property");
+            reportProgress("Done Migrating from Relation to Property", statements.size(), statements.size());
+        }
+    }
+
+    /**
+     * @since 11.1
+     */
+    protected void migrateSessionPropertyToSecured(CoreSession session, MigrationContext migrationContext) {
+        List<DocumentModel> comments = getUnsecuredComments(session);
+        PropertyCommentManager commentManager = (PropertyCommentManager) Framework.getService(CommentManager.class);
+        processBatched(migrationContext, BATCH_SIZE, comments,
+                comment -> migrateCommentsFromPropertyToSecured(session, commentManager, comment),
+                "Migrating comments from Property to Secured");
+
+        String queryCommentsRootFolder = String.format("SELECT %s FROM Document WHERE %s='%s' AND %s='%s'", ECM_UUID,
+                ECM_NAME, COMMENTS_DIRECTORY, ECM_PRIMARYTYPE, HIDDEN_FOLDER_TYPE);
+        session.query(queryCommentsRootFolder).forEach((folder -> session.removeDocument(folder.getRef())));
```

Comment:
Use the `removeDocuments` API that takes all the refs at once (and can optimize the removal in some cases — or could in the future).